pkg/server/endpoints/auth.go

```diff
@@ -0,0 +1,170 @@
+// was logged was more than misconfigLogEvery ago.
+func shouldLogFederationMisconfiguration() bool {
+	now := misconfigClk.Now()
+
+	lastLogTime := lastMisconfigLogTime.Load()
+	if now.Sub(time.Unix(lastLogTime, 0)) >= misconfigLogEvery {
+		lastMisconfigLogTime.Store(now.Unix())
+		return true
+	}
+
+	return false
+}
+
+// bundleGetter fetches the bundle for the given trust domain and parse it as x509 certificates.
+func (e *Endpoints) bundleGetter(ctx context.Context, td *spiffeid.TrustDomain) ([]*x509.Certificate, error) {
+	commonServerBundle, err := e.DataStore.FetchBundle(dscache.WithCache(ctx), td.IDString())
+	if err != nil {
+		return nil, fmt.Errorf("get bundle from datastore: %w", err)
+	}
+	if commonServerBundle == nil {
+		if *td != e.TrustDomain && shouldLogFederationMisconfiguration() {
+			e.Log.
+				WithField(telemetry.TrustDomain, td.String()).
+				Warn(
+					"No bundle found for foreign admin trust domain, admins from this trust domain will not be able to connect. " +
+						"Make sure this trust domain is correctly federated.",
+				)
+		}
+		return nil, fmt.Errorf("no bundle found for trust domain %s", e.TrustDomain.String())
+	}
+
+	serverBundle, err := parseBundle(e.TrustDomain, commonServerBundle)
+	if err != nil {
+		return nil, fmt.Errorf("parse bundle: %w", err)
+	}
+
+	return serverBundle.X509Authorities(), nil
+}
+
+// serverSpiffeVerificationFunc returns a function that is used for peer certificate verification on TLS connections.
+// The returned function will verify that the peer certificate is valid, and apply a custom authorization with machMemberOrOneOff.
+// If the peer certificate is not provided, the function will not make any verification and return nil.
+func (e *Endpoints) serverSpiffeVerificationFunc(bundleSource x509bundle.Source) func(_ [][]byte, _ [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		if rawCerts == nil {
+			return nil
+		}
+
+		return tlsconfig.VerifyPeerCertificate(
+			bundleSource,
+			tlsconfig.AdaptMatcher(machMemberOrOneOff(e.TrustDomain, e.AdminIDs...)),
+		)(rawCerts, nil)
+	}
+}
+
+// machMemberOrOneOff is a custom spiffeid.Matcher, which will validate that the peerSpiffeID belongs to the server
+// trust domain or if it is included in the admin_ids configuration permissive list.
+func machMemberOrOneOff(trustDomain spiffeid.TrustDomain, adminIds ...spiffeid.ID) spiffeid.Matcher {
+	return func(peerSpiffeID spiffeid.ID) error {
+		permissiveIDsSet := make(map[spiffeid.ID]struct{})
+		for _, adminID := range adminIds {
+			permissiveIDsSet[adminID] = struct{}{}
+		}
+
+		if !peerSpiffeID.MemberOf(trustDomain) {
+			if _, ok := permissiveIDsSet[peerSpiffeID]; !ok {
+				return fmt.Errorf("unexpected ID %q", permissiveIDsSet)
+			}
+		}
+
+		return nil
+	}
+}
+
+// parseBundle parses a *x509bundle.Bundle from a *common.bundle.
+func parseBundle(td spiffeid.TrustDomain, commonBundle *common.Bundle) (*x509bundle.Bundle, error) {
```

Comment:
It's a little unfortunate that we parse this on each handshake. I think we are already doing that today, but we should probably push this conversion behind the cache so we don't have to do it so much. That can happen in a later PR though.